src/MKRWAN.h

```diff
@@ -917,6 +947,10 @@ class LoRaModem : public Stream
     if (waitResponse(timeout, "+EVENT=1,1") != 1) {
       return false;
     }
+    // Second answer, always returns +OK
+    if (waitResponse() != 1) {
+      return false;
+    }
```

Comment:
Hi @flhofer, I tested you PRs. In ABP everything works fine, but not in OTAA. 
When `modem.joinOTAA(appEui, appKey)` is run in the sketch, the connection actually takes place and I am able to see it in the TTN console. However, these 3 additional lines of code return a failure for OTAA. The result is that even if the OTAA connection happened, the sketch shows a failure.
Removing these 3 lines, I am able to successfully connect both in ABP and OTAA.

Reply:
Hi @giulcioffi 
Which firmware version are you using? Recent versions (>1.2.0) respond with additional `+OK`s. I had the opposite problem where it was waiting forever for a `=` when checking status due to an additional `+OK` without =

Reply:
We might need to decide here. `+OK` everywhere as it is implemented now does not allow discerning if the `+OK` answer is the response to a command or to a query `+OK=`.
My proposal is to change responses to command:value, like I wrote in the other pull request for the firmware.

Reply:
I used the firmware from your PR https://github.com/arduino/mkrwan1300-fw/pull/35 :) This issue is encountered only in OTAA. If you could test OTAA and eventually modify the firmware to provide that additional +OK, it would be great!

Reply:
Hi @giulcioffi,
I finally figured where the `+OK` came from and removed it altogether (a library thing issuing two commands in a row). While there, I abstracted getters and setters to common calls, simplifying the library and easing the change to `+<CMD>=<VALUE>` responses. 

This new push now contains the following changes
- added timeout to `skipUntil`
- two common getters for almost all calls
- one setter for almost all calls
- an optimization of waitResponse that
    1. peeks the buffer and checks match only on delimiter hit ('\r','=','+')
    2. allows for parsing of multiple responses in the same string without compromising buffer, e.g. `+EVENT=0,0+OK\r`
    3. parses binary receive `+RECVB=`

- getter response parse match, e.g. `+DR?` responds with `+DR=5` and autodetect based on the version number to accept `+OK=5` too

I tested it for a few hours here, but I will continue tomorrow. Anyone who can join is warmly welcome

Florian

Reply:
Hi @flhofer, thanks for your effort!! I compiled the firmware with your latest commit https://github.com/arduino/mkrwan1300-fw/pull/35/commits/f44055655d0cabc5764dc6121d26f4c221d9040c, but I am then unable to start the module. Is this still a work in progress or am I missing something?

Reply:
@giulcioffi yes, still somehow work in progress. For my project, I need almost everything in the library to work seamlessly but I got stuck with the `+RFPOWER` command not working. Anyway, at least now with the latest changes I get a proper response `+ERR_PARAM`, telling me where to look.

This to say, I shared the changes but did not test everything thoroughly. Anyway, I agree let's keep it here and fix everything until there is a certain amount of confidence that it works as expected.

Reply:
@giulcioffi fixed the `start module` issue. It was an inverted return value check in the `begin`, the only setter in the whole code like this 😆 

In the meantime, I found also some other small issues in the firmware like the TXpower set that didn't work. Also, some of the library functions aren't even implemented in the firmware. I added the constants in the library, but obviously, the modem won't do anything.

#define AT_UART		  "+UART"
#define AT_FACNEW	  "+FACNEW"
#define AT_SLEEP	  "+SLEEP"
#define AT_MSIZE	  "+MSIZE"

More tests and updates to come

Reply:
Hi @flhofer, thank you again for your contribution! I just tested the new firmware and library in OTAA and ABP and the communication works perfectly :) please keep us updated on your tests so that we can then have a final review.

Reply:
I think this part is done :)